<!DOCTYPE HTML>
<html>
<head>
    <title>protein-visualizer</title>
    <link rel="stylesheet" href="styles.css">
    <link rel="icon" href="favicon.ico">
    <meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
    <nav>
        <a href="#Home"><b>protein-visualizer</b><img src="favicon.ico" id="Icon"/></a>
        <a href="#Usage">Usage</a>
        <a href="#Code">Code</a>
        <a href="#Attribution">Attribution</a>
    </nav>
    <section id="Content">
        <section id="Home">
            <h1>protein-visualizer</h1>
            <p>A free and open-source program to visualize proteins and the function of its individual amino acids
                through deep learning.</p>
            <img src="demo_run.png" alt="Demonstration of running the visualizer"/>
            <h2>How does it work?</h2>
            <p>This interactive visualizer uses a neural network to identify functional and structural clusters within
            a protein by calculating embeddings, numerical expressions of each amino acid. The high-dimensional
            data is then transformed into 2D in order to easily interpret the possible functional components of a
            protein. Next, a graph convolutional network uses both the 3D structure and amino acid sequence to predict
            Gene Ontology (GO) annotations that describe the molecular function of the protein. Through
            Gradient-weighted Class Activation Mapping (GradCAM), the program can calculate how much each amino
            acid contributed to each GO prediction and subsequently allow for the identification of functional residues.</p>
            <p>This program was written entirely with Python, using Pyglet as an OpenGL interface.</p>
            <p><b>Disclaimer:</b> All of the data (except for 3D structure) shown in this program is generated through machine
            learning and thus cannot be verified to be completely accurate. It is recommended to double-check
                the predicted functions through experimental sources </p>

        </section>
        <section id="Usage">
            <h1>Usage</h1>
            <h2>Installation</h2>
            <p>Download and extract the latest release from the
                <a href="https://github.com/kiwijuice56/protein-visualizer/releases">protein-visualizer release page.</a>
            Run <code>protein-visualizer-x.x/main.exe</code> and select your protein file in the file dialog. If your file
                has many chains, you can select one by typing its ID in the program terminal when prompted.
            </p>
            <p>The visualizer accepts a Protein Data Bank (PDB) format file encoding the structure and sequence of
                a protein. You can download proteins from
                <a href="https://www.rcsb.org/">RCSB Protein Data Bank</a> to test out with this program.
            </p>
            <p>Before viewing a protein, the program caches amino acid embeddings and Gene Ontology (GO) annotations in
                <code>protein-visualizer-x.x/data/[protein_name]_data.json</code> to reduce load times the next time the
                program is run with the selected protein.
            </p>
            <p>If you wish to use the prediction data yourself, you can parse through this file.
                The following information is stored as a dictionary for each protein:
            </p>
            <ul>
                <li><code>GO_ids</code>: A list of predicted GO IDs.</li>
                <li><code>GO_names</code>: A list of descriptions for each predicted GO ID.</li>
                <li><code>confidence</code>: A number between 0 and 1 denoting the confidence of each GO prediction.</li>
                <li><code>saliency_maps</code>: A matrix where each row contains numbers between 0 and 1 denoting how
                    strongly an amino acid contributed to the prediction of that row's GO annotation.</li>
                <li><code>sequence</code>: The protein's amino acid sequence in the FASTA format (without a header).</li>
                <li><code>embedding_points</code>: A flattened list of (x, y) points denoting the coordinates of each
                    amino acid in the latent space after being transformed into 2D.</li>
                <li><code>cluster_indices</code>: A list of integers denoting which cluster each amino acid belongs to, with -1 meaning no cluster.</li>
            </ul>
            <p>You can also generate the data file without opening the GUI by executing the program from the command line:
            <code>protein-visualizer.x.x/main.exe [path_to_pdb_file] [optional_chain_id]</code>.</p>
            <p> The <code>protein-visualizer-x.x/saved_models/</code> folder contains the pre-trained models created by
            the ProSE and DeepFRI teams.</p>

            <h2>Controls</h2>
            <ul>
                <li><code>Left Mouse</code>: Translate the camera and navigate the embedding space.</li>
                <li><code>Right/Middle Mouse</code>: Rotate the camera.</li>
                <li><code>Up/Down Arrow</code>: Increase/decrease atom point size.</li>
                <li><code>O</code>: Toggles atom outline.</li>
            </ul>
        </section>
        <section id="Code">
            <h1>Code</h1>
            <p>All of the code is available under the MIT license on the <a href="https://github.com/kiwijuice56/protein-visualizer">protein-visualizer GitHub repository.</a>
            Feel free to take any part of the code to expand or transform, just make sure you keep the attributions to ProSE and DeepFRI!
        </section>
        <section id="Attribution">
            <h1>Attribution</h1>
            <ul>
                <li>
                    <a href="https://ericalfaro.dev/">Eric Alfaro</a>: Primary program developer as part of an MIT UROP project under the guidance of Dr. Manolis Kellis and Dr. Tianlong Chen.
                </li>
                <li>
                    <p><a href="https://github.com/tbepler/prose/">Protein Sequence Embeddings (ProSE)</a>: Amino acid embedding code and pre-trained models.</p>
                    <ul>
                        <li>Bepler, T., Berger, B. Learning the protein language: evolution, structure, and function. Cell Systems 12, 6 (2021). https://doi.org/10.1016/j.cels.2021.05.017</li>
                        <li>Bepler, T., Berger, B. Learning protein sequence embeddings using information from structure. International Conference on Learning Representations (2019). https://openreview.net/pdf?id=SygLehCqtm</li>
                    </ul>
                </li>
                <li>
                    <p><a href="https://github.com/flatironinstitute/DeepFRI">Deep Functional Residue Identification (DeepFRI)</a>: GO annotation prediction code and pre-trained models.</p>
                    <ul>
                        <li>Gligorijevic, Vladimir and Renfrew, P. Douglas and Kosciolek, Tomasz and Leman, Julia Koehler and Cho, Kyunghyun and Vatanen, Tommi and Berenberg, Daniel and Taylor, Bryn and Fisk, Ian M. and Xavier, Ramnik J. and Knight, Rob and Bonneau, Richard. Structure-based function prediction using graph convolutional networks. Cold Spring Harbor Laboratory (2019). https://doi.org/10.1101/786236.</li>
                    </ul>
                </li>
            </ul>
        </section>
        <footer>Â©2024 Eric Alfaro </footer>
    </section>
</body>

</html>